File: android/app/src/main/java/com/mattermost/helpers/RealPathUtil.java
package com.mattermost.helpers;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.provider.OpenableColumns;
import android.content.ContentResolver;
import android.os.Environment;
import android.webkit.MimeTypeMap;
import android.util.Log;
import android.text.TextUtils;

import android.os.ParcelFileDescriptor;

import java.io.*;
import java.nio.channels.FileChannel;

// Class based on DocumentHelper https://gist.github.com/steveevers/a5af24c226f44bb8fdc3

public class RealPathUtil {
    public static final String CACHE_DIR_NAME = "mmShare";
    public static String getRealPathFromURI(final Context context, final Uri uri) {

        // DocumentProvider
        if (DocumentsContract.isDocumentUri(context, uri)) {
            // ExternalStorageProvider
            if (isExternalStorageDocument(uri)) {
                final String docId = DocumentsContract.getDocumentId(uri);
                final String[] split = docId.split(":");
                final String type = split[0];

                if ("primary".equalsIgnoreCase(type)) {
                    return Environment.getExternalStorageDirectory() + "/" + split[1];
                }
            } else if (isDownloadsDocument(uri)) {
                // DownloadsProvider

                final String id = DocumentsContract.getDocumentId(uri);
                if (!TextUtils.isEmpty(id)) {
                    if (id.startsWith("raw:")) {
                        return id.replaceFirst("raw:", "");
                    }
                    try {
                        return getPathFromSavingTempFile(context, uri);
                    } catch (NumberFormatException e) {
                        Log.e("ReactNative", "DownloadsProvider unexpected uri " + uri);
                        return null;
                    }
                }
            } else if (isMediaDocument(uri)) {
                // MediaProvider

                final String docId = DocumentsContract.getDocumentId(uri);
                final String[] split = docId.split(":");
                final String type = split[0];

                Uri contentUri = null;
                if ("image".equals(type)) {
                    contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
                } else if ("video".equals(type)) {
                    contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
                } else if ("audio".equals(type)) {
                    contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
                }

                final String selection = "_id=?";
                final String[] selectionArgs = new String[] {
                        split[1]
                };

                String name = getDataColumn(context, contentUri, selection, selectionArgs);
                if (!TextUtils.isEmpty(name)) {
                    return name;
                }

                return getPathFromSavingTempFile(context, uri);
            }
        }

        if ("content".equalsIgnoreCase(uri.getScheme())) {
            // MediaStore (and general)

            if (isGooglePhotosUri(uri)) {
                return uri.getLastPathSegment();
            }

            // Try save to tmp file, and return tmp file path
            return getPathFromSavingTempFile(context, uri);
        } else if ("file".equalsIgnoreCase(uri.getScheme())) {
            return uri.getPath();
        }

        return null;
    }

    public static String getPathFromSavingTempFile(Context context, final Uri uri) {
        File tmpFile;
        String fileName = "";

        if (uri == null || uri.isRelative()) {
            return null;
        }

        // Try and get the filename from the Uri
        try {
            Cursor returnCursor =
                    context.getContentResolver().query(uri, null, null, null, null);
            int nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
            returnCursor.moveToFirst();
            fileName = sanitizeFilename(returnCursor.getString(nameIndex));
            returnCursor.close();
        } catch (Exception e) {
            // just continue to get the filename with the last segment of the path
        }

        try {
            if (TextUtils.isEmpty(fileName)) {
                fileName = sanitizeFilename(uri.getLastPathSegment().trim());
            }


            File cacheDir = new File(context.getCacheDir(), CACHE_DIR_NAME);
            boolean cacheDirExists = cacheDir.exists();
            if (!cacheDirExists) {
                cacheDirExists = cacheDir.mkdirs();
            }

            if (cacheDirExists) {
                tmpFile = new File(cacheDir, fileName);
                boolean fileCreated = tmpFile.createNewFile();

                if (fileCreated) {
                    ParcelFileDescriptor pfd = context.getContentResolver().openFileDescriptor(uri, "r");

                    try (FileInputStream inputSrc = new FileInputStream(pfd.getFileDescriptor())) {
                        FileChannel src = inputSrc.getChannel();
                        try (FileOutputStream outputDst = new FileOutputStream(tmpFile)) {
                            FileChannel dst = outputDst.getChannel();
                            dst.transferFrom(src, 0, src.size());
                            src.close();
                            dst.close();
                        }
                    }

                    pfd.close();
                }
                return tmpFile.getAbsolutePath();
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        return null;
    }

    public static String getDataColumn(Context context, Uri uri, String selection,
                                       String[] selectionArgs) {

        Cursor cursor = null;
        final String column = "_data";
        final String[] projection = {
                column
        };

        try {
            cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs,
                    null);
            if (cursor != null && cursor.moveToFirst()) {
                final int index = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(index);
            }
        } finally {
            if (cursor != null)
                cursor.close();
        }
        return null;
    }


    public static boolean isExternalStorageDocument(Uri uri) {
        return "com.android.externalstorage.documents".equals(uri.getAuthority());
    }

    public static boolean isDownloadsDocument(Uri uri) {
        return "com.android.providers.downloads.documents".equals(uri.getAuthority());
    }

    public static boolean isMediaDocument(Uri uri) {
        return "com.android.providers.media.documents".equals(uri.getAuthority());
    }

    public static boolean isGooglePhotosUri(Uri uri) {
        return "com.google.android.apps.photos.content".equals(uri.getAuthority());
    }

    public static String getExtension(String uri) {
        String extension = "";
        if (uri == null) {
            return extension;
        }

        extension = MimeTypeMap.getFileExtensionFromUrl(uri);
        if (!extension.equals("")) {
            return extension;
        }

        int dot = uri.lastIndexOf(".");
        if (dot >= 0) {
            return uri.substring(dot);
        } else {
            // No extension.
            return "";
        }
    }

    public static String getMimeType(File file) {

        String extension = getExtension(file.getName());

        if (extension.length() > 0)
            return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.substring(1));

        return "application/octet-stream";
    }

    public static String getMimeType(String filePath) {
        File file = new File(filePath);
        return getMimeType(file);
    }

     public static String getMimeTypeFromUri(final Context context, final Uri uri) {
        try {
            ContentResolver cR = context.getContentResolver();
            return cR.getType(uri);
        } catch (Exception e) {
            return "application/octet-stream";
        }
    }

    public static void deleteTempFiles(final File dir) {
        try {
            if (dir.isDirectory()) {
                deleteRecursive(dir);
            }
        } catch (Exception e) {
            // do nothing
        }
    }

    private static void deleteRecursive(File fileOrDirectory) {
        if (fileOrDirectory.isDirectory()) {
                File[] files = fileOrDirectory.listFiles();
                if (files != null) {
                    for (File child : files)
                        deleteRecursive(child);
                }
        }

        if (!fileOrDirectory.delete()) {
            Log.i("ReactNative", "Couldn't delete file " + fileOrDirectory.getName());
        }
    }

    private static String sanitizeFilename(String filename) {
        if (filename == null) {
            return null;
        }

        File f = new File(filename);
        return f.getName();
    }
}

File: libraries/@mattermost/rnutils/android/src/main/java/com/mattermost/rnutils/helpers/RealPathUtil.kt
package com.mattermost.rnutils.helpers

import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.provider.DocumentsContract
import android.provider.MediaStore
import android.provider.OpenableColumns
import android.text.TextUtils
import android.util.Log
import android.webkit.MimeTypeMap
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

const val CACHE_DIR_NAME: String = "mmShare"

object RealPathUtil {
    init {
        deleteTempFiles(File(CACHE_DIR_NAME))
    }

    fun getRealPathFromURI(context:Context, uri:Uri?): String? {
        if (uri == null) {
            return null
        }
        // DocumentProvider
        if (DocumentsContract.isDocumentUri(context, uri)) {
            // ExternalStorageProvider
            if (isExternalStorageDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val type = split[0]
                if ("primary".equals(type, ignoreCase = true)) {
                    return context.getExternalFilesDir(split[1])?.absolutePath
                }
            } else if (isDownloadsDocument(uri)) {
                // DownloadsProvider
                val id = DocumentsContract.getDocumentId(uri)
                if (!TextUtils.isEmpty(id)) {
                    if (id.startsWith("raw:")) {
                        return id.replaceFirst(("raw:").toRegex(), "")
                    }

                    try {
                        return getPathFromSavingTempFile(context, uri)
                    } catch (e:NumberFormatException) {
                        Log.e("ReactNative", "DownloadsProvider unexpected uri $uri")
                        return null
                    }
                }
            } else if (isMediaDocument(uri)) {
                // MediaProvider
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val type = split[0]
                var contentUri: Uri? = null
                when (type) {
                    "image" -> {
                        contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                    }
                    "video" -> {
                        contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                    }
                    "audio" -> {
                        contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
                    }
                }
                val selectionArgs = arrayOf(split[1])
                return contentUri?.let { getDataColumn(context, it, selectionArgs) }
            }
        }
        if ("content".equals(uri.scheme, ignoreCase = true)) {
            // MediaStore (and general)
            if (isGooglePhotosUri(uri)) {
                return uri.lastPathSegment
            }
            // Try save to tmp file, and return tmp file path
            return getPathFromSavingTempFile(context, uri)
        }
        else if ("file".equals(uri.scheme, ignoreCase = true)) {
            return uri.path
        }
        return null
    }

    private fun getPathFromSavingTempFile(context:Context, uri:Uri): String? {
        val tmpFile:File
        var fileName: String? = null
        // Try and get the filename from the Uri
        try
        {
            val returnCursor = context.contentResolver.query(uri, null, null, null, null)
            val nameIndex = returnCursor?.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            returnCursor?.moveToFirst()
            fileName = sanitizeFilename(nameIndex?.let { returnCursor.getString(it) })
            returnCursor?.close()
        } catch (e:Exception) {
            // just continue to get the filename with the last segment of the path
        }

        try
        {
            if (fileName == null) {
                fileName = sanitizeFilename(uri.lastPathSegment.toString().trim())
            }
            val cacheDir = File(context.cacheDir, CACHE_DIR_NAME)
            if (!cacheDir.exists()) {
                cacheDir.mkdirs()
            }
            tmpFile = File(cacheDir, fileName!!)
            tmpFile.createNewFile()
            val pfd = context.contentResolver.openFileDescriptor(uri, "r")
            val src = FileInputStream(pfd?.fileDescriptor).channel
            val dst = FileOutputStream(tmpFile).channel
            dst.transferFrom(src, 0, src.size())
            src.close()
            dst.close()
            pfd?.close()
        }
        catch (ex:IOException) {
            return null
        }
        return tmpFile.absolutePath
    }

    private fun sanitizeFilename(filename: String?): String? {
        if (filename == null) {
            return null
        }

        val f = File(filename)
        return f.name
    }

    private fun getDataColumn(context:Context, uri:Uri, selectionArgs:Array<String>): String? {
        var cursor: Cursor? = null
        val column = "_data"
        val selection = "_id=?"
        val projection = arrayOf(column)
        try
        {
            cursor = context.contentResolver.query(uri, projection, selection, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val index = cursor.getColumnIndexOrThrow(column)
                return cursor.getString(index)
            }
        }
        finally
        {
            cursor?.close()
        }

        return null
    }

    private fun isExternalStorageDocument(uri:Uri):Boolean {
        return "com.android.externalstorage.documents" == uri.authority
    }

    private fun isDownloadsDocument(uri:Uri):Boolean {
        return "com.android.providers.downloads.documents" == uri.authority
    }

    private fun isMediaDocument(uri:Uri):Boolean {
        return "com.android.providers.media.documents" == uri.authority
    }

    private fun isGooglePhotosUri(uri:Uri):Boolean {
        return "com.google.android.apps.photos.content" == uri.authority
    }

    private fun getExtension(uri: String?): String? {
        if (uri == null) {
            return null
        }

        val dot = uri.lastIndexOf(".")
        return if (dot >= 0)
        {
            uri.substring(dot)
        } else {
            // No extension.
            ""
        }
    }

    private fun getMimeType(file: File): String? {
        val extension = getExtension(file.name)
        if (extension?.length!! > 0) {
            return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.substring(1))
        }

        return "application/octet-stream"
    }

    fun getMimeType(filePath: String): String? {
        val file = File(filePath)
        return getMimeType(file)
    }

    private fun deleteTempFiles(dir:File) {
        try
        {
            if (dir.isDirectory) {
                deleteRecursive(dir)
            }
        }
        catch (e:Exception) {
            // do nothing
        }
    }

    private fun deleteRecursive(fileOrDirectory:File) {
        if (fileOrDirectory.isDirectory) {
            val files = fileOrDirectory.listFiles()
            if (files?.size!! > 0) {
                for (child in files) {
                    deleteRecursive(child)
                }
            }
            fileOrDirectory.delete()
        }
    }
}
